src/asami/query.cljc

```diff
@@ -323,23 +323,49 @@
        part)
       col-meta)))
 
+(defn index
+  {:private true}
+  [xs]
+  (zipmap xs (range)))
+
 (s/defn disjunction
   "Implements an OR operation by repeating a join across each arm of the operation,
    and concatenating the results"
   [graph
    part :- Results
-   [_ & patterns]]  ;; Discard the first element, since it is just the OR operator
+   [_ & patterns]] ;; Discard the first element, since it is just the OR operator
   (let [spread (map #(left-join % part graph) patterns)
-        cols (:cols (meta (first spread)))]
-    (doseq [s (rest spread)]
-      (when (not= (:cols (meta s)) cols)
-        (throw (ex-info
-                "Alternate sides of OR clauses may not contain different vars"
-                (zipmap patterns (map (comp :cols meta) spread))))))
+        ;; Column information of each sequence of results.
+        spread-cols (into [] (map (comp :cols meta)) spread)
+        ;; Combine all distinct column information from left to right
+        ;; to produce new column information for the result set.
+        result-cols (into [] (comp cat (distinct)) spread-cols)]
     (with-meta
       ;; Does distinct create a scaling issue?
-      (*select-distinct* (sequence cat spread))
-      {:cols cols})))
+      (*select-distinct*
+       ;; For each result sequence `results` of `spread` with columns
+       ;; that are not equal to result-cols, reorganize the elements
+       ;; of `results` such that they align with `result-cols`.
+       (sequence 
+        (comp (map (fn [results cols]
+                     (if (= cols result-cols)
```

Comment:
Instead of running this inside `map`, the comparison can be done outside. When `(= cols result-cols)` then the result can be:
`(sequence cat spread)`
This will be cheaper and faster when the variables match on both sides

Reply:
Not sure I understand, there is no `cols` outside this `map`?  Before we removed it in this PR, `cols` was bound to `(:cols (meta (first spread)))`, is that what you're referring to? I'm not sure how we can short-circuit decision-making about our output by inspecting only `(first spread)`, but I might be missing something. :)

Reply:
Yeah, my comment is messed up. Sorry. Not sure what I was thinking exactly.

However, the thing I was trying to get to was the simple form of disjunctions where each argument has the same set of columns. So something like on line 346:
```clojure
(if (apply = (map (comp cols :meta) spread))
  (do
    (assert (= result-cols (-> spread cols :meta)))  ;; not actually needed
    (apply concat spread))
  ;; other side of if:
  ;; For each result sequence...
```

Otherwise, it all looks good

Reply:
OK. I've looked again, and realized that I was misreading what this map does.